File: tests/oracle/callspec_noreturn_1204.cc
/*
 * CALLSPEC-NORETURN-1204: locked Ghidra 12.0.4 oracle for the FuncProto /
 * FuncCallSpecs no-return flag lifecycle (CALLSPEC-NORETURN-WIRE-0001
 * slice a).
 *
 * Covered projection (all observations go through public Ghidra API):
 *  - FuncProto::FuncProto (fspec.cc:3778): flags == 0, so isNoReturn() and
 *    isInline() default to false; FuncCallSpecs(PcodeOp*) (fspec.cc:4924)
 *    inherits the same false defaults through FuncProto().
 *  - setNoReturn (fspec.hh:1439) / isNoReturn (fspec.hh:1434): explicit set,
 *    repeated-set idempotence, clear via false — driven both on a plain
 *    FuncProto and through the FuncCallSpecs inherited surface (the surface
 *    flow.cc:747 truncateIndirectJump uses).
 *  - copyFlowEffects (fspec.cc:3806-3812): only the is_inline|no_return
 *    subset is copied, as a one-way overwrite (clear-then-OR: a source bit
 *    of 0 clears the destination bit); modellock and everything else stays
 *    untouched. This is the flow.cc:664 queryCall channel that propagates a
 *    callee's noreturn state onto the call site.
 *  - FuncProto::copy (fspec.cc:3789) copies the flags wholesale;
 *    FuncCallSpecs::clone (fspec.cc:4964-4977) preserves the copied
 *    FuncProto bits, rebinds to the new op, and resets the active-input
 *    recovery state.
 *  - No void->noreturn inference exists anywhere in the fspec layer: a
 *    void-returning prototype (setInternal + a decoded <prototype> with a
 *    void <returnsym> and no noreturn attribute) stays isNoReturn()==false.
 *    The only producers of the bit are the explicit XML attribute
 *    (fspec.cc:4720-4722), OptionNoReturn (options.cc:358), a direct
 *    setNoReturn, and copyFlowEffects/copy propagation.
 *  - decode/encode channel (fspec.cc:4675-4840 / 4625-4667): the
 *    noreturn="true" attribute sets the bit, an absent attribute leaves it
 *    clear, noreturn="false" leaves it clear, and encode emits the
 *    attribute only when the bit is set.
 *
 * The fixture architecture (spaces + TypeFactory + a decoded "fixture"
 * ProtoModel) mirrors the action_merge_order_1204 fixture; the two bare
 * PcodeOps carry the CALLIND opcode so the FuncCallSpecs constructor leaves
 * the entry address invalid exactly like a CALLIND call site.
 */

#include <bits/stdc++.h>

// Test-only access observer for PcodeOp::setOpcode (op.cc:276), the same
// #define pattern the fspec_phase0 / address_space fixtures use; layout is
// unchanged (access specifiers only), and the mangled setOpcode symbol
// resolves against the locked archive. <bits/stdc++.h> is included first so
// the defines never leak into libstdc++ headers.
#define class struct
#define private public
#define protected public
#include "op.hh"
#undef protected
#undef private
#undef class

#include "architecture.hh"
#include "database.hh"
#include "fspec.hh"
#include "libdecomp.hh"
#include "marshal.hh"
#include "translate.hh"
#include "type.hh"

using namespace ghidra;

class FixtureTranslate final : public Translate {
  VarnodeData dummy_register;

public:
  FixtureTranslate() {
    setBigEndian(false);
    setUniqueBase(0);
    insertSpace(new ConstantSpace(this, this));
    insertSpace(new AddrSpace(this, this, IPTR_PROCESSOR, "other", false, 8, 1, 1,
                              AddrSpace::hasphysical, 0, 0));
    insertSpace(new UniqueSpace(this, this, 2, 0));
    AddrSpace *ram = new AddrSpace(this, this, IPTR_PROCESSOR, "ram", false,
                                   8, 1, 3, AddrSpace::hasphysical, 0, 0);
    insertSpace(ram);
    AddrSpace *reg = new AddrSpace(this, this, IPTR_PROCESSOR, "register", false,
                                   8, 1, 4, AddrSpace::hasphysical, 0, 0);
    insertSpace(reg);
    SpacebaseSpace *stack = new SpacebaseSpace(
        this, this, "stack", 5, 8, ram, 1, true);
    insertSpace(stack);
    VarnodeData stack_pointer;
    stack_pointer.space = reg;
    stack_pointer.offset = 0;
    stack_pointer.size = 8;
    addSpacebasePointer(stack, stack_pointer, 8, true);
    insertSpace(new AddrSpace(this, this, IPTR_PROCESSOR, "join", false, 8, 1, 6,
                              AddrSpace::hasphysical, 0, 0));
    insertSpace(new IopSpace(this, this, 7));
    setDefaultCodeSpace(3);
    dummy_register.space = getSpace(4);
    dummy_register.offset = 0;
    dummy_register.size = 8;
  }

  void initialize(DocumentStorage &) override {}
  const VarnodeData &getRegister(const std::string &) const override {
    return dummy_register;
  }
  std::string getRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  std::string getExactRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  void getAllRegisters(std::map<VarnodeData, std::string> &) const override {}
  void getUserOpNames(std::vector<std::string> &) const override {}
  int4 instructionLength(const Address &) const override { return 0; }
  int4 oneInstruction(PcodeEmit &, const Address &) const override { return 0; }
  int4 printAssembly(AssemblyEmit &, const Address &) const override { return 0; }
};

class FixtureArchitecture final : public Architecture {
protected:
  Translate *buildTranslator(DocumentStorage &) override { return (Translate *)0; }
  void buildLoader(DocumentStorage &) override {}
  PcodeInjectLibrary *buildPcodeInjectLibrary(void) override {
    return (PcodeInjectLibrary *)0;
  }
  void buildTypegrp(DocumentStorage &) override {}
  void buildCoreTypes(DocumentStorage &) override {}
  void buildCommentDB(DocumentStorage &) override {}
  void buildStringManager(DocumentStorage &) override {}
  void buildConstantPool(DocumentStorage &) override {}
  void buildContext(DocumentStorage &) override {}
  void buildSymbols(DocumentStorage &) override {}
  void buildSpecFile(DocumentStorage &) override {}
  void modifySpaces(Translate *) override {}
  void resolveArchitecture(void) override {}

public:
  FixtureArchitecture() {
    FixtureTranslate *fixture_translate = new FixtureTranslate();
    translate = fixture_translate;
    copySpaces(fixture_translate);
    max_basetype_size = 16;
    types = new TypeFactory(this);
    types->setupSizes();
    types->setCoreType("xunknown1", 1, TYPE_UNKNOWN, false);
    types->setCoreType("xunknown2", 2, TYPE_UNKNOWN, false);
    types->setCoreType("xunknown4", 4, TYPE_UNKNOWN, false);
    types->setCoreType("xunknown8", 8, TYPE_UNKNOWN, false);
    types->cacheCoreTypes();
    TypeOp::registerInstructions(inst, types, translate);
    symboltab = new Database(this, false);
    symboltab->attachScope(new ScopeInternal(0x101, "", this), (Scope *)0);
    ProtoModel *model = new ProtoModel(this);
    std::istringstream stream(
        "<prototype name=\"fixture\" extrapop=\"0\"><input/><output/></prototype>");
    XmlDecode decoder(this);
    decoder.ingestStream(stream);
    model->decode(decoder);
    protoModels[model->getName()] = model;
    setDefaultModel(model);
  }

  ProtoModel *fixture_model(void) { return protoModels["fixture"]; }
  TypeOp *callind_behavior(void) { return inst[CPUI_CALLIND]; }

  void printMessage(const std::string &) const override {}
};

// A bare CALLIND-opcoded PcodeOp: the FuncCallSpecs constructor takes the
// non-CALL branch and leaves the entry address invalid (fspec.cc:4942).
static PcodeOp *make_callind_op(FixtureArchitecture &arch, uintm ord) {
  PcodeOp *op = new PcodeOp(1, SeqNum(Address(), ord));
  op->setOpcode(arch.callind_behavior());
  return op;
}

// Decode a <prototype> element into a FuncProto backed by the fixture
// model + void type. Mirrors the Funcdata prototype-restore staging
// (setInternal before decode, fspec.cc:4679-4680).
static void decode_proto(FixtureArchitecture &arch, FuncProto &proto,
                         const std::string &xml) {
  proto.setInternal(arch.fixture_model(), arch.types->getTypeVoid());
  std::istringstream stream(xml);
  XmlDecode decoder(&arch);
  decoder.ingestStream(stream);
  proto.decode(decoder, &arch);
}

int main(void) {
  std::cout << std::unitbuf;
  std::vector<std::string> spec_paths;
  startDecompilerLibrary(spec_paths);
  std::cout << "schema=1|fixture=CALLSPEC-NORETURN-1204|"
               "oracle=e40ed13014025f82488b1f8f7bca566894ac376b\n";
  try {
    FixtureArchitecture arch;
    TypeVoid void_type;

    // ---- case=default_ctor ------------------------------------------
    {
      FuncProto proto;
      proto.setInternal((ProtoModel *)0, &void_type);
      PcodeOp *op = make_callind_op(arch, 1);
      FuncCallSpecs fc(op);
      std::cout << "case=default_ctor"
                << "|proto_noret=" << (int4)proto.isNoReturn()
                << "|proto_inline=" << (int4)proto.isInline()
                << "|callspec_noret=" << (int4)fc.isNoReturn()
                << "|callspec_inline=" << (int4)fc.isInline()
                << "|callspec_entry_invalid="
                << (int4)fc.getEntryAddress().isInvalid()
                << '\n';
    }

    // ---- case=explicit_set_idempotent ------------------------------
    {
      FuncProto proto;
      proto.setInternal((ProtoModel *)0, &void_type);
      proto.setNoReturn(true);
      int4 a = proto.isNoReturn();
      proto.setNoReturn(true);
      int4 b = proto.isNoReturn();
      proto.setNoReturn(false);
      int4 c = proto.isNoReturn();
      proto.setNoReturn(false);
      int4 d = proto.isNoReturn();
      // Through the FuncCallSpecs inherited surface (flow.cc:747 uses
      // fc->setNoReturn(true) on a callspec).
      PcodeOp *op = make_callind_op(arch, 2);
      FuncCallSpecs fc(op);
      fc.setNoReturn(true);
      int4 e = fc.isNoReturn();
      fc.setNoReturn(false);
      int4 f = fc.isNoReturn();
      std::cout << "case=explicit_set_idempotent"
                << "|set_true=" << a
                << "|set_true_again=" << b
                << "|set_false=" << c
                << "|set_false_again=" << d
                << "|callspec_set_true=" << e
                << "|callspec_clear=" << f
                << '\n';
    }

    // ---- case=copy_flow_effects -------------------------------------
    {
      // callee{inline=T,noret=T} -> fc{F,F} becomes {T,T}
      FuncProto callee1;
      callee1.setInternal((ProtoModel *)0, &void_type);
      callee1.setInline(true);
      callee1.setNoReturn(true);
      PcodeOp *op = make_callind_op(arch, 3);
      FuncCallSpecs fc(op);
      fc.copyFlowEffects(callee1);
      int4 s1_noret = fc.isNoReturn();
      int4 s1_inline = fc.isInline();

      // callee{F,F} -> fc{previous T,T} clears to {F,F} (one-way overwrite)
      FuncProto callee2;
      callee2.setInternal((ProtoModel *)0, &void_type);
      fc.copyFlowEffects(callee2);
      int4 s2_noret = fc.isNoReturn();
      int4 s2_inline = fc.isInline();

      // callee{inline=F,noret=T} -> fc {F,T}; modellock (a non-subset flag
      // on the source) must not leak into the callspec.
      FuncProto callee3;
      callee3.setInternal((ProtoModel *)0, &void_type);
      callee3.setNoReturn(true);
      callee3.setModelLock(true);
      fc.copyFlowEffects(callee3);
      int4 s3_noret = fc.isNoReturn();
      int4 s3_inline = fc.isInline();
      int4 ml_leak = fc.isModelLocked();
      int4 ml_src = callee3.isModelLocked();
      std::cout << "case=copy_flow_effects"
                << "|step1_noret=" << s1_noret
                << "|step1_inline=" << s1_inline
                << "|step2_noret=" << s2_noret
                << "|step2_inline=" << s2_inline
                << "|step3_noret=" << s3_noret
                << "|step3_inline=" << s3_inline
                << "|modellock_not_copied=" << (int4)(ml_leak == 0)
                << "|src_modellock_stays=" << ml_src
                << '\n';
    }

    // ---- case=full_copy_and_clone -----------------------------------
    {
      FuncProto a;
      a.setInternal((ProtoModel *)0, &void_type);
      a.setNoReturn(true);
      a.setInline(true);
      FuncProto b;
      b.setInternal((ProtoModel *)0, &void_type);
      b.copy(a);
      int4 copy_noret = b.isNoReturn();
      int4 copy_inline = b.isInline();

      PcodeOp *op1 = make_callind_op(arch, 4);
      PcodeOp *op2 = make_callind_op(arch, 5);
      FuncCallSpecs *fc1 = new FuncCallSpecs(op1);
      // initActiveInput dereferences the prototype model (fspec.cc:5335
      // getMaxInputDelay), so stage the fixture model first.
      fc1->setInternal(arch.fixture_model(), arch.types->getTypeVoid());
      fc1->setNoReturn(true);
      fc1->setInline(true);
      fc1->initActiveInput();
      FuncCallSpecs *clone = fc1->clone(op2);
      int4 clone_noret = clone->isNoReturn();
      int4 clone_inline = clone->isInline();
      int4 clone_rebound = (clone->getOp() == op2) && (op1 != op2);
      int4 clone_active_reset =
          (clone->isInputActive() == false && fc1->isInputActive() == true);
      std::cout << "case=full_copy_and_clone"
                << "|copy_noret=" << copy_noret
                << "|copy_inline=" << copy_inline
                << "|clone_noret=" << clone_noret
                << "|clone_inline=" << clone_inline
                << "|clone_rebound=" << clone_rebound
                << "|clone_active_reset=" << (int4)clone_active_reset
                << '\n';
      delete clone;
      delete fc1;
      delete op2;
      delete op1;
    }

    // ---- case=void_no_inference -------------------------------------
    {
      // A void-returning prototype never gains noreturn implicitly.
      FuncProto proto;
      proto.setInternal(arch.fixture_model(), arch.types->getTypeVoid());
      int4 ctor_void = proto.isNoReturn();
      FuncProto decoded;
      decode_proto(
          arch, decoded,
          "<prototype model=\"fixture\" extrapop=\"0\">"
          "<returnsym typelock=\"true\">"
          "<addr space=\"ram\" offset=\"0x0\" size=\"1\"/>"
          "<void/>"
          "</returnsym></prototype>");
      int4 decode_void = decoded.isNoReturn();
      int4 decode_void_output = decoded.getOutputType()->getMetatype() == TYPE_VOID;
      std::cout << "case=void_no_inference"
                << "|setinternal_void_noret=" << ctor_void
                << "|decode_void_noret=" << decode_void
                << "|decode_output_is_void=" << (int4)decode_void_output
                << '\n';
    }

    // ---- case=decode_encode_channel ----------------------------------
    {
      FuncProto with_true;
      decode_proto(
          arch, with_true,
          "<prototype model=\"fixture\" extrapop=\"0\" noreturn=\"true\">"
          "<returnsym typelock=\"true\">"
          "<addr space=\"ram\" offset=\"0x0\" size=\"1\"/>"
          "<void/>"
          "</returnsym></prototype>");
      int4 d_true = with_true.isNoReturn();

      FuncProto with_absent;
      decode_proto(
          arch, with_absent,
          "<prototype model=\"fixture\" extrapop=\"0\">"
          "<returnsym typelock=\"true\">"
          "<addr space=\"ram\" offset=\"0x0\" size=\"1\"/>"
          "<void/>"
          "</returnsym></prototype>");
      int4 d_absent = with_absent.isNoReturn();

      FuncProto with_false;
      decode_proto(
          arch, with_false,
          "<prototype model=\"fixture\" extrapop=\"0\" noreturn=\"false\">"
          "<returnsym typelock=\"true\">"
          "<addr space=\"ram\" offset=\"0x0\" size=\"1\"/>"
          "<void/>"
          "</returnsym></prototype>");
      int4 d_false = with_false.isNoReturn();

      // encode emits the noreturn attribute only when the bit is set.
      std::ostringstream out_set;
      XmlEncode enc_set(out_set);
      with_true.encode(enc_set);
      int4 e_present = out_set.str().find("noreturn=") != std::string::npos;
      std::ostringstream out_clear;
      XmlEncode enc_clear(out_clear);
      with_absent.encode(enc_clear);
      int4 e_absent = out_clear.str().find("noreturn=") != std::string::npos;
      std::cout << "case=decode_encode_channel"
                << "|decode_true=" << d_true
                << "|decode_absent=" << d_absent
                << "|decode_false_attr=" << d_false
                << "|encode_present=" << e_present
                << "|encode_absent=" << e_absent
                << '\n';
    }
  } catch (const LowlevelError &err) {
    std::cerr << "fixture LowlevelError: " << err.explain << '\n';
    return 1;
  }
  return 0;
}

File: tests/oracle/callspec_noreturn_1204.rs
//! CALLSPEC-NORETURN-1204: Rust comparand for the locked Ghidra 12.0.4
//! oracle `callspec_noreturn_1204.cc` (CALLSPEC-NORETURN-WIRE-0001 slice a).
//!
//! Byte-mirrors the C++ projection of the FuncProto/FuncCallSpecs no-return
//! flag lifecycle: constructor defaults (flags==0), explicit
//! set/clear/idempotence through both the FuncProto and the FuncCallSpecs
//! inherited surface, copyFlowEffects' is_inline|no_return one-way subset
//! overwrite (flow.cc:664 queryCall channel), full copy + callspec clone
//! propagation with op rebind and active-input reset, the absence of any
//! void->noreturn inference, and the decode/encode attribute channel.
//!
//! The C++ side drives the real Ghidra classes against a fixture
//! architecture (spaces + TypeFactory + decoded "fixture" ProtoModel); this
//! side mirrors each observation through the crate's public fspec API plus
//! the marshal TreeDecoder/TreeEncoder, so both stdouts are byte-identical.

use std::sync::{Arc, RwLock};

use rugra::address::{Address, SeqNum};
use rugra::fspec::{FuncCallSpecs, FuncProto};
use rugra::marshal::{xml_tree, AttributeId, Decoder, ElementId, IdRegistry, TreeDecoder, TreeEncoder};
use rugra::op::{PcodeOp, PcodeOpRef};
use rugra::opcodes::OpCode;
use rugra::type_system::datatype::{Datatype, TypeBase, TypeMetatype};

// RUGRA-GLUE: fixture-local canonical void type, the same construction the
// funcproto_lock_1204 comparand uses (C++ reads arch.types->getTypeVoid()).
fn void_type() -> Arc<Datatype> {
    Arc::new(Datatype::Void(TypeBase::new(
        "void".to_string(),
        0,
        TypeMetatype::Void,
    )))
}

// A bare CALLIND-opcoded op: FuncCallSpecs::new_for_op only reads input(0)
// for CPUI_CALL, so the spec's entry stays None exactly like the C++
// constructor's non-CALL branch (fspec.cc:4942).
fn make_callind_op(ord: u32) -> PcodeOpRef {
    PcodeOpRef(Arc::new(RwLock::new(PcodeOp::new(
        SeqNum::new(Address::new(0), ord),
        OpCode::CPUI_CALLIND,
    ))))
}

// Mirror of the C++ decode_proto staging: setInternal before decode
// (fspec.cc:4679-4680). The closures hand-roll the minimal <returnsym>
// child reader the flat FuncProto decode needs (C++ resolves storage
// through ProtoStoreInternal inside FuncProto::decode).
fn decode_proto(xml: &str) -> FuncProto {
    let mut proto = FuncProto::new(String::new(), void_type());
    proto.set_internal(None, void_type());
    let doc = xml_tree(xml.as_bytes()).expect("fixture xml parses");
    let ids = Arc::new(RwLock::new(IdRegistry::new()));
    let mut dec = TreeDecoder::from_document(&doc, ids);
    let model_resolver = |name: &str| name == "fixture";
    let decode_output_storage =
        |dec: &mut dyn Decoder| -> (Address, Arc<Datatype>, bool) {
            // <returnsym typelock=".."><addr ../><void/></returnsym>
            let returnsym = dec.open_element();
            let mut output_lock = false;
            loop {
                let aid = dec.next_attribute_id();
                if aid == 0 {
                    break;
                }
                match dec.attribute_name(aid).as_deref() {
                    Some("typelock") => output_lock = dec.read_bool(),
                    _ => {
                        let _ = dec.read_string();
                    }
                }
            }
            let mut addr = Address::new(0);
            if dec.peek_element() != 0 {
                let addr_id = dec.open_element();
                loop {
                    let aid = dec.next_attribute_id();
                    if aid == 0 {
                        break;
                    }
                    match dec.attribute_name(aid).as_deref() {
                        Some("offset") => {
                            addr = Address::new(dec.read_string().parse::<u64>().unwrap_or(0));
                        }
                        _ => {
                            let _ = dec.read_string();
                        }
                    }
                }
                dec.close_element(addr_id);
            }
            let mut ty = void_type();
            if dec.peek_element() != 0 {
                let type_id = dec.open_element();
                // C++ TypeFactory::decodeTypeNoRef (type.cc:4436): the
                // ELEM_VOID child (<void/>) resolves to the void type.
                let type_name = dec
                    .element_name(type_id)
                    .unwrap_or_default()
                    .to_string();
                loop {
                    let aid = dec.next_attribute_id();
                    if aid == 0 {
                        break;
                    }
                    let _ = dec.read_string();
                }
                if type_name != "void" {
                    panic!("fixture only decodes the void type, got {type_name}");
                }
                dec.close_element(type_id);
                ty = void_type();
            }
            dec.close_element(returnsym);
            (addr, ty, output_lock)
        };
    proto
        .decode(&mut dec, &model_resolver, &decode_output_storage)
        .expect("FuncProto::decode succeeds");
    proto
}

// Mirror of FuncProto::encode through the TreeEncoder, checking the
// noreturn attribute presence the way the C++ side greps its XML stream.
fn encode_has_noreturn(proto: &FuncProto) -> bool {
    let registry = Arc::new(RwLock::new(IdRegistry::new()));
    let mut enc = TreeEncoder::new(registry);
    let prototype_elem = ElementId::new("prototype", 169);
    let model_attrib = AttributeId::new("model", 13);
    let extrapop_attrib = AttributeId::new("extrapop", 6);
    let dotdotdot_attrib = AttributeId::new("dotdotdot", 115);
    let modellock_attrib = AttributeId::new("modellock", 122);
    let inline_attrib = AttributeId::new("inline", 118);
    let noreturn_attrib = AttributeId::new("noreturn", 123);
    let constructor_attrib = AttributeId::new("constructor", 4);
    let destructor_attrib = AttributeId::new("destructor", 5);
    let returnsym_elem = ElementId::new("returnsym", 172);
    let typelock_attrib = AttributeId::new("typelock", 23);
    let unaffected_elem = ElementId::new("unaffected", 173);
    let killedbycall_elem = ElementId::new("killedbycall", 162);
    let returnaddress_elem = ElementId::new("returnaddress", 5);
    let likelytrash_elem = ElementId::new("likelytrash", 163);
    let addr_elem = ElementId::new("addr", 11);
    let space_attrib = AttributeId::new("space", 20);
    let offset_attrib = AttributeId::new("offset", 16);
    let size_attrib = AttributeId::new("size", 19);
    proto.encode(
        &mut enc,
        &prototype_elem,
        &model_attrib,
        &extrapop_attrib,
        &dotdotdot_attrib,
        &modellock_attrib,
        &inline_attrib,
        &noreturn_attrib,
        &constructor_attrib,
        &destructor_attrib,
        &returnsym_elem,
        &typelock_attrib,
        &unaffected_elem,
        &killedbycall_elem,
        &returnaddress_elem,
        &likelytrash_elem,
        &addr_elem,
        &space_attrib,
        &offset_attrib,
        &size_attrib,
        rugra::space::AddressSpace::Ram,
        0,
        1,
        &[],
        &[],
        &|_, _, _| rugra::fspec::EffectType::UnknownEffect,
        &|_| {},
    );
    let doc = enc.into_document();
    let root = doc.root.as_ref().unwrap();
    let root = root.read().unwrap();
    root.attr_names.iter().any(|name| name == "noreturn")
}

fn main() {
    println!(
        "schema=1|fixture=CALLSPEC-NORETURN-1204|\
         oracle=e40ed13014025f82488b1f8f7bca566894ac376b"
    );

    // ---- case=default_ctor ------------------------------------------
    {
        let mut proto = FuncProto::new(String::new(), void_type());
        proto.set_internal(None, void_type());
        let op = make_callind_op(1);
        let fc = FuncCallSpecs::new_for_op(&op, FuncProto::new(String::new(), void_type()));
        println!(
            "case=default_ctor|proto_noret={}|proto_inline={}|callspec_noret={}|callspec_inline={}|callspec_entry_invalid={}",
            u8::from(proto.is_no_return()),
            u8::from(proto.is_inline()),
            u8::from(fc.is_no_return()),
            u8::from(fc.is_inline()),
            u8::from(fc.entry_addr.is_none()),
        );
    }

    // ---- case=explicit_set_idempotent ------------------------------
    {
        let mut proto = FuncProto::new(String::new(), void_type());
        proto.set_internal(None, void_type());
        proto.set_no_return(true);
        let a = u8::from(proto.is_no_return());
        proto.set_no_return(true);
        let b = u8::from(proto.is_no_return());
        proto.set_no_return(false);
        let c = u8::from(proto.is_no_return());
        proto.set_no_return(false);
        let d = u8::from(proto.is_no_return());
        // Through the FuncCallSpecs inherited surface (flow.cc:747 uses
        // fc->setNoReturn(true) on a callspec).
        let op = make_callind_op(2);
        let mut fc = FuncCallSpecs::new_for_op(&op, FuncProto::new(String::new(), void_type()));
        fc.set_no_return(true);
        let e = u8::from(fc.is_no_return());
        fc.set_no_return(false);
        let f = u8::from(fc.is_no_return());
        println!(
            "case=explicit_set_idempotent|set_true={a}|set_true_again={b}|set_false={c}|set_false_again={d}|callspec_set_true={e}|callspec_clear={f}",
        );
    }

    // ---- case=copy_flow_effects -------------------------------------
    {
        // callee{inline=T,noret=T} -> fc{F,F} becomes {T,T}
        let mut callee1 = FuncProto::new(String::new(), void_type());
        callee1.set_internal(None, void_type());
        callee1.set_inline(true);
        callee1.set_no_return(true);
        let op = make_callind_op(3);
        let mut fc = FuncCallSpecs::new_for_op(&op, FuncProto::new(String::new(), void_type()));
        fc.copy_flow_effects(&callee1);
        let (s1_noret, s1_inline) = (u8::from(fc.is_no_return()), u8::from(fc.is_inline()));

        // callee{F,F} -> fc{previous T,T} clears to {F,F} (one-way overwrite)
        let mut callee2 = FuncProto::new(String::new(), void_type());
        callee2.set_internal(None, void_type());
        fc.copy_flow_effects(&callee2);
        let (s2_noret, s2_inline) = (u8::from(fc.is_no_return()), u8::from(fc.is_inline()));

        // callee{inline=F,noret=T} -> fc {F,T}; modellock (a non-subset flag
        // on the source) must not leak into the callspec.
        let mut callee3 = FuncProto::new(String::new(), void_type());
        callee3.set_internal(None, void_type());
        callee3.set_no_return(true);
        callee3.set_model_lock(true);
        fc.copy_flow_effects(&callee3);
        let (s3_noret, s3_inline) = (u8::from(fc.is_no_return()), u8::from(fc.is_inline()));
        let ml_leak = u8::from(fc.prototype.is_model_locked());
        let ml_src = u8::from(callee3.is_model_locked());
        println!(
            "case=copy_flow_effects|step1_noret={s1_noret}|step1_inline={s1_inline}|step2_noret={s2_noret}|step2_inline={s2_inline}|step3_noret={s3_noret}|step3_inline={s3_inline}|modellock_not_copied={}|src_modellock_stays={ml_src}",
            u8::from(ml_leak == 0),
        );
    }

    // ---- case=full_copy_and_clone -----------------------------------
    {
        let mut a = FuncProto::new(String::new(), void_type());
        a.set_internal(None, void_type());
        a.set_no_return(true);
        a.set_inline(true);
        let mut b = FuncProto::new(String::new(), void_type());
        b.set_internal(None, void_type());
        b.copy_from(&a);
        let copy_noret = u8::from(b.is_no_return());
        let copy_inline = u8::from(b.is_inline());

        let op1 = make_callind_op(4);
        let op2 = make_callind_op(5);
        let mut fc1 = FuncCallSpecs::new_for_op(&op1, FuncProto::new(String::new(), void_type()));
        fc1.set_no_return(true);
        fc1.set_inline(true);
        fc1.init_active_input();
        let clone = fc1.clone_for_op(&op2);
        let clone_noret = u8::from(clone.is_no_return());
        let clone_inline = u8::from(clone.is_inline());
        let clone_rebound = u8::from(
            clone
                .op
                .upgrade()
                .map(|c| Arc::ptr_eq(&c, &op2.0))
                .unwrap_or(false)
                && !Arc::ptr_eq(&op1.0, &op2.0),
        );
        let clone_active_reset =
            u8::from(!clone.is_input_active() && fc1.is_input_active());
        println!(
            "case=full_copy_and_clone|copy_noret={copy_noret}|copy_inline={copy_inline}|clone_noret={clone_noret}|clone_inline={clone_inline}|clone_rebound={clone_rebound}|clone_active_reset={clone_active_reset}",
        );
    }

    // ---- case=void_no_inference -------------------------------------
    {
        // A void-returning prototype never gains noreturn implicitly.
        let mut proto = FuncProto::new(String::new(), void_type());
        proto.set_internal(None, void_type());
        let ctor_void = u8::from(proto.is_no_return());
        let decoded = decode_proto(
            "<prototype model=\"fixture\" extrapop=\"0\">\
             <returnsym typelock=\"true\">\
             <addr space=\"ram\" offset=\"0x0\" size=\"1\"/>\
             <void/>\
             </returnsym></prototype>",
        );
        let decode_void = u8::from(decoded.is_no_return());
        let decode_void_output = u8::from(matches!(
            *decoded.return_type,
            Datatype::Void(_)
        ));
        println!(
            "case=void_no_inference|setinternal_void_noret={ctor_void}|decode_void_noret={decode_void}|decode_output_is_void={decode_void_output}",
        );
    }

    // ---- case=decode_encode_channel ----------------------------------
    {
        let with_true = decode_proto(
            "<prototype model=\"fixture\" extrapop=\"0\" noreturn=\"true\">\
             <returnsym typelock=\"true\">\
             <addr space=\"ram\" offset=\"0x0\" size=\"1\"/>\
             <void/>\
             </returnsym></prototype>",
        );
        let d_true = u8::from(with_true.is_no_return());

        let with_absent = decode_proto(
            "<prototype model=\"fixture\" extrapop=\"0\">\
             <returnsym typelock=\"true\">\
             <addr space=\"ram\" offset=\"0x0\" size=\"1\"/>\
             <void/>\
             </returnsym></prototype>",
        );
        let d_absent = u8::from(with_absent.is_no_return());

        let with_false = decode_proto(
            "<prototype model=\"fixture\" extrapop=\"0\" noreturn=\"false\">\
             <returnsym typelock=\"true\">\
             <addr space=\"ram\" offset=\"0x0\" size=\"1\"/>\
             <void/>\
             </returnsym></prototype>",
        );
        let d_false = u8::from(with_false.is_no_return());

        // encode emits the noreturn attribute only when the bit is set.
        let e_present = u8::from(encode_has_noreturn(&with_true));
        let e_absent = u8::from(encode_has_noreturn(&with_absent));
        println!(
            "case=decode_encode_channel|decode_true={d_true}|decode_absent={d_absent}|decode_false_attr={d_false}|encode_present={e_present}|encode_absent={e_absent}",
        );
    }
}
